### PR TITLE
Require primary IP addresses for ready hosts

### PR DESCRIPTION
**Pull request by mpolden**
This disallows a host from moving to ready until it has IP addresses. A scenario
where this can happen:

- A host is dynamically provisioned
- The host sets itself ready before
  `DynamicProvisioningMaintainer#resumeProvisioning` writes the assigned IPs
- The host ends up in reserved without IP addresses, unable to activate (because `active` requires IPs)
- `DynamicProvisioningMaintainer#resumeProvisioning` no longer considers the
  host because it's not in `provisioned`
  
@hakonhall

FYI @freva

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java [hunk 1]
```diff
@@ -97,11 +97,16 @@ public Node(String id, IP.Config ipConfig, String hostname, Optional<String> par
...
         if (parentHostname.isPresent()) {
-            if (!ipConfig.pool().getIpSet().isEmpty()) throw new IllegalArgumentException("A child node cannot have an IP address pool");
+            if (!ipConfig.pool().isEmpty()) throw new IllegalArgumentException("A child node cannot have an IP address pool");
             if (modelName.isPresent()) throw new IllegalArgumentException("A child node cannot have model name set");
             if (switchHostname.isPresent()) throw new IllegalArgumentException("A child node cannot have switch hostname set");
         }
 
+        if (type.isHost() && state == State.ready) {
+            requireNonEmpty(ipConfig.primary(), "A " + type + " must have at least one primary IP address in state " + state);
+            if (type == NodeType.host) requireNonEmpty(ipConfig.pool().asSet(), "A " + type + " must have a non-empty IP address pool in state " + state);
```
Comment (hakonhall) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java:107]: I think all host types require non-empty IP address pool.

**Review by freva — APPROVED** [review]

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/IP.java [hunk 2]
```diff
@@ -325,10 +325,14 @@ public int eventuallyUnusedAddressCount(NodeList nodes) {
             return ipAddresses.protocol;
         }
 
-        public Set<String> getIpSet() {
+        public Set<String> asSet() {
```
Comment (hakonhall) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/IP.java:328]: A Pool is a pool of IP addresses and/or hostnames, which is why this method should be named getIpSet().

**Comment by freva** [comment]
Oops, I'm just FYI, didn't mean to approve. Leaving this for @hakonhall

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/IP.java [hunk 3]
```diff
@@ -325,10 +325,14 @@ public int eventuallyUnusedAddressCount(NodeList nodes) {
...
             return ipAddresses.asSet();
         }
 
+        public boolean isEmpty() {
```
Comment (hakonhall) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/IP.java:332]: The method name is misleading as it indicates there are no addresses in this pool either.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java (hunk 1 above)
Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java:107]: Agree, but many tests violate this. Ok, I'll fix it...

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/IP.java (hunk 3 above)
Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/IP.java:332]: Ok, I'll revert. It's very confusing that an _IP address pool_ contains something that's not an IP address, which is also called "Address".

**Comment by mpolden** [comment]
PTAL

**Review by hakonhall — APPROVED** [review]

**Pull request merged**